File: runtime/go/units/wasi/fd_seek.go
// requires: memory/i64_store
func (w *WASI) wasi_fd_seek(fd uint32, offset uint64, whence, outPtr uint32) uint32 {
    f, ok := w.fds[fd].(*os.File)
    if !ok {
        return wasiBadf
    }
    if w.isStdio(f) {
        return wasiSpipe
    }
    // WASI whence (0=set, 1=cur, 2=end) matches Go's io.Seek* values exactly.
    if whence > 2 {
        return wasiInval
    }
    if e := w.checkRight(fd, rightFdSeek); e != wasiOk {
        return e
    }
    pos, err := f.Seek(int64(offset), int(whence))
    if err != nil {
        // A negative resulting offset is EINVAL, not an I/O error (the suite
        // distinguishes them).
        return wasiInval
    }
    w.memory.i64_store(uint64(outPtr), uint64(pos))
    return wasiOk
}

File: runtime/java/units/wasi/fd_seek.java
// requires: memory/i64_store
// WASI whence: 0=set, 1=cur, 2=end. FileChannel has no whence, so compute the
// absolute position. Stdio is not seekable (SPIPE).
int wasi_fd_seek(int fd, long offset, int whence, int outPtr) {
    Object e = fds.get(fd);
    if (isStdio(e)) {
        return WASI_SPIPE;
    }
    if (!(e instanceof Handle)) {
        return WASI_BADF;
    }
    java.nio.channels.FileChannel ch = ((Handle) e).ch;
    try {
        long pos;
        switch (whence) {
            case 0:
                pos = offset;
                break;
            case 1:
                pos = ch.position() + offset;
                break;
            case 2:
                pos = ch.size() + offset;
                break;
            default:
                return WASI_INVAL;
        }
        ch.position(pos);
        memory.i64_store(Integer.toUnsignedLong(outPtr), pos);
    } catch (java.io.IOException ex) {
        return WASI_IO;
    }
    return WASI_OK;
}
